pkg/kritis/admission/admission_test.go

```diff
@@ -78,12 +78,10 @@ func Test_UnqualifiedImage(t *testing.T) {
 	mockISP := func(namespace string) ([]kritisv1beta1.ImageSecurityPolicy, error) {
 		return []kritisv1beta1.ImageSecurityPolicy{{}}, nil
 	}
-	mockMetadata := func() (metadata.MetadataFetcher, error) {
-		return mockMetadataClient{}, nil
-	}
+
 	mockConfig := config{
 		retrievePod:                 mockPod,
-		fetchMetadataClient:         mockMetadata,
+		fetchMetadataClient:         testutil.EmptyMockMetadata(),
```

Comment:
A few testing readability comments here:

- I'd rather see the mock metadata inline than moved into a helper method. When writing tests, one should not sacrifice clarity in the name of de-duplication.
- "Empty" is a misnomer, as the helper does in fact setup attestation and severity.
- I can't tell if it's the case here, but tests should never rely on default values specified by a helper method, since that forces readers to read the helper method’s implementation details in order to understand the test.

Reply:
hmm, the reason why i moved it to helper is because, you need to do a bunch of mock setups, instead of focusing on the values you want to test your code against.
The idea of these mocks is users can just pass the vulnerabilites and attestations that image has.
say, i want to write a test case where my-isp allows max severity LOW and the image has a HIGH severity vuln, i can use the helper to do so, 
```
vulnzHigh = metadata.Vulnerability{
		CVE:             "cve2",
		Severity:        "HIGH",
		HasFixAvailable: false,
	}
```
And setup the mock to pass this value using.
```
mockMetadataClient = testutil.MockMetadataClient{
		Vulnz: []metadata.Vulnerability{
			vulnz1,
		},
		PGPAttestations: nil,
	}
// test my code.
```

VS create a MetadataFetcher interface in you tests and implementing all the required methods:
```
type mockMetadataClient struct {
}

func (m mockMetadataClient) GetVulnerabilities(containerImage string) ([]metadata.Vulnerability, error) {
	return []metadata.Vulnerability{
		vulnHigh,
	}, nil
}

func (m mockMetadataClient) CreateAttestationOccurence(note *containeranalysispb.Note,
	containerImage string,
	pgpSigningKey *secrets.PGPSigningSecret) (*containeranalysispb.Occurrence, error) {
	return nil, nil
}

func (m mockMetadataClient) GetAttestationNote(aa kritisv1beta1.AttestationAuthority) (*containeranalysispb.Note, error) {
	return nil, nil
}

func (m mockMetadataClient) GetAttestations(containerImage string) ([]metadata.PGPAttestation, error) {
	return <some attestations>, nil
}

// test my code.
```
Doing this for every test case, would be too cumbersome. You can declare `mockMetadataClient`  struct which accepts vuln and pgp attestations and return those. You will have to repeat that code in every test file. The helper just pulls it out and provides a public interface which you can use everywhere.

Does that make sense?

Regarding EmptyMockMetadata: EmptyMockMetadata() returns no vulnerabilities and attestations. Maybe i shd i rename it to `NilMetadataFethcer.`

Reply:
Seems reasonable.

In tests, repeated code is far more preferable to de-duplicated code, but there is a healthy balance to be made. Tests should be optimized for readability, not writability, since readers will always approach tests with an air of skepticism, and authors will approach them with an air of confidence.